internal/querycoord/handoff_handler.go

```diff
@@ -0,0 +1,386 @@
+func (handler *HandoffHandler) process(segmentID int64) error {
+	task := handler.tasks[segmentID]
+	// if task is cancel and success, clean up
+	switch task.state {
+	case handoffTaskCancel, handoffTaskDone:
+		overrideSegments := handler.getOverrideSegments(task)
+		for _, daughterSegmentID := range overrideSegments {
+			handler.clean(daughterSegmentID)
+		}
+		handler.clean(segmentID)
+
+	case handoffTaskInit:
+		isValid, collectionInfo := handler.verifyRequest(task.segmentInfo)
+		if !isValid || !Params.QueryCoordCfg.AutoHandoff {
+			task.state = handoffTaskCancel
+			log.Info("HandoffHandler: collection/partition has not been loaded, task canceled", zap.Int64("segmentID", task.segmentInfo.SegmentID))
+			return ErrHandoffRequestInvalid
+		}
+		// TODO, add segment lock here, if segment lock add failed,
+		if !task.locked {
+			if err := handler.broker.acquireSegmentsReferLock(handler.ctx, task.segmentInfo.SegmentID, []UniqueID{task.segmentInfo.SegmentID}); err != nil {
+				// if task can not be holded, there are three possible situation
+				// 1. temporary fail of data coord -> retry
+				// 2. collection is dropped -> verify Handoff Req valid should find the collection is released, the verifyHandoffReqValid before lock should handle it.
+				// 3. compaction happened -> we should soon received another handoff task to handle current task.
+				log.Warn("HandoffHandler: acquire segment reference lock failed", zap.Int64("segmentID", task.segmentInfo.SegmentID), zap.Error(err))
+				return fmt.Errorf("failed to acquire segment refer lock")
+			}
+		}
+		task.locked = true
+		// TODO we should not directly poll the index info, wait for notification should be a better idea.
+		indexInfo, err := handler.broker.getIndexInfo(handler.ctx, task.segmentInfo.CollectionID, task.segmentInfo.SegmentID, collectionInfo.Schema)
+		if err == nil {
+			// if index exist or not enableIndex, ready to load
+			task.segmentInfo.IndexInfos = indexInfo
+			// NOTICE, this is the trick, if compaction happened recursively, it should be all in our task list.
+			task := &HandOffTask{
+				task.segmentInfo, handoffTaskReady, true,
+			}
+			handler.tasks[task.segmentInfo.SegmentID] = task
+			handler.notify <- false
+			log.Info("HandoffHandler: enqueue indexed segments", zap.Int64("segmentID", task.segmentInfo.SegmentID))
+		}
+
+	case handoffTaskReady:
+		validHandoffReq, _ := handler.verifyRequest(task.segmentInfo)
+		if !validHandoffReq || !Params.QueryCoordCfg.AutoHandoff {
+			task.state = handoffTaskCancel
+			log.Info("HandoffHandler: collection/partition has not been loaded, task canceled", zap.Int64("segmentID", task.segmentInfo.SegmentID))
+			return ErrHandoffRequestInvalid
+		}
+		handler.triggerHandoff(task)
+	}
+	// handoffTaskTriggered state don't need to be handled in the loop, it will handled by the go routine in triggerHandoff
+
+	return nil
+}
+
+func (handler *HandoffHandler) clean(segmentID int64) {
+	task := handler.tasks[segmentID]
```

Comment:
should we clean ChangeInfoMetaPrefix here?

Reply:
> should we clean ChangeInfoMetaPrefix here?

QueryNode cleans it, a new shard leader that NodeDown balanced to, can't capture it if we clean it here